actionpack/lib/action_dispatch/middleware/cookies.rb

```diff
@@ -560,63 +573,98 @@ def serializer
         def digest
           request.cookies_digest || "SHA1"
         end
-
-        def key_generator
-          request.key_generator
-        end
     end
 
-    class SignedCookieJar < AbstractCookieJar # :nodoc:
+    class SignedKeyRotatingCookieJar < AbstractCookieJar # :nodoc:
       include SerializedCookieJars
 
       def initialize(parent_jar)
         super
-        secret = key_generator.generate_key(request.signed_cookie_salt)
-        @verifier = ActiveSupport::MessageVerifier.new(secret, digest: digest, serializer: ActiveSupport::MessageEncryptor::NullSerializer)
+
+        secret = key_generator.generate_key(signed_cookie_salt)
+        @verifier = ActiveSupport::MessageVerifier.new(secret, digest: signed_cookie_digest, serializer: SERIALIZER)
+
+        Array.wrap(signed_cookie_rotate_out).each do |outbound|
+          @verifier.rotate(*expand_outbound_rotation(outbound))
```

Comment:
I think your `expand_outbound_rotation` is doing too many things. Can it be simplified to something like?

```ruby
Array.wrap(signed_cookie_rotate_out).each do |digest:, **rotation|
  @verifier.rotate(extract_key(rotation), digest: digest, serializer: SERIALIZER)
end

def extract_key(raw_key: nil, secret: nil, salt: nil)
  if raw_key
    raw_key
  else
    raise ArgumentError, "Missing secret and salt to generate key from" unless secret && salt

    ActiveSupport::KeyGenerator.new(secret, iterations: 1000).generate_key(salt)
  end
end
```

Reply:
I dig it :+1: